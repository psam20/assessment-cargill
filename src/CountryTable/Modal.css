.modal {
    position: fixed;
    top: 0;
    left: 0;
    width:100%;
    height: 100%;
    background-color: rgb(71, 156, 212);
  }
  .modal-main {
    position:fixed;
    background-image: linear-gradient(to right, rgba(236, 168, 168, 0), rgb(159, 226, 114));
    width: 50%;
    height: auto;
    align-items: center;
    top:50%;
    left:50%;
    transform: translate(-50%,-50%);
  }
  
  .display-block {
    display: block;
  }
  
  .display-none {
    display: none;
  }