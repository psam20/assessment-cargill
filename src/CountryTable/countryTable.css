
body{
        margin: 0;
        background-image: linear-gradient(to right, rgba(236, 168, 168, 0), rgb(59, 132, 243));
}
.header{
        background-color:#398B93;
        color:white;
        font-size: 1.5em;
        padding: 1rem;
        text-align: center;
        text-transform: uppercase;
     }
     .country-table{
        border: 1px solid #398B93;
        border-radius: 10px;
        box-shadow: 3px 3px 0 rgba(0,0,0,0.1);
        max-width: 90%;
        margin: 1em auto;
        overflow: hidden;
        width: 800px;
     }
     .table{
             width: 100%;
     }
     .td ,.btn,.th{
             color: #398B93;
             padding: 10px;
     }
     .btn:hover{
             cursor: pointer;
     }
     .tdh:hover{
             cursor: pointer;
     }
     .td{
             text-align: center;
             vertical-align: middle;
     }
     .td,.th:first-child{
        font-size: 0.95em;
        line-height: 1.4;
        text-align: left;
     }
     .th{
                background-color: rgb(143, 143, 224);
                text-align: left;
                font-weight: 300;
             
     }
     .tr:nth-child(2n){
             background-color: white;
     }
     .tr:nth-child(2n+1){
             background-color:#e4e0e0;
     }