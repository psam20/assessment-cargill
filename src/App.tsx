import React, { useEffect } from 'react';
import CountryTable from './CountryTable/countryTable';

import './app.css';
export interface props {
 name:string,
 
}

function App() {

  const [result, setResult] = React.useState<Array<object>>([]);
  const [loading,setLoading]=React.useState<boolean>(false);
  const [data,setData]=React.useState<Array<object>>(result);
  const [searchText, setSearchText] = React.useState<string>("");
  

   useEffect(() => {
     setLoading(true)
    fetch('https://restcountries.eu/rest/v2/all')
      .then(res => res.json())
      .then(res => {
        setResult(res)
        setLoading(false)
      })
      .catch(err=>console.log(err))
  }, [])

 
 

  useEffect(()=>{
       const value=searchText;
       const filterProperty=["name","region","alpha2Code","alpha3Code"]
      const lowerCaseValue=value.toLowerCase().trim();
      if (lowerCaseValue===''){
        setData(result)
        console.log(result)
      }
      else{
        const filteredData=result.filter((item:any)=> {
           
          return Object.keys(item).some((key:any)=>
            filterProperty.includes(key)? item[key].toString().toLowerCase().includes(lowerCaseValue) :false)
        })
  
        setData(filteredData)
        
       console.log(filteredData);
      }
  
     
      

  },[searchText,result])
  
 
  if(loading){
    return <h1>Loading Countries Data...</h1>
  }

  return (
    <div>
              <div className="filter-div">
                <input  className="input" type="text" value={searchText} placeholder="Search" onChange={(e)=>setSearchText(e.target.value)} name="search-filter" />
              </div>
        <CountryTable Countries={data} />
        
    </div>
  );
}

export default App;
