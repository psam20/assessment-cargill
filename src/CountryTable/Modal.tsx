import React from 'react';
import './Modal.css';
interface props {
    modalData: any,
    languages:any,
    currency:any,
    show: boolean,
    handleClose: VoidFunction,
    
    
}

const Modal = (props: props) => {

    
 

    const showHideClassName = props.show ? 'modal display-block' : 'modal display-none';
    return (<div className={showHideClassName}>
        <section className='modal-main'>
            <div style={{ alignItems: "center" }}>
                <img src={`${props.modalData.flag}`} alt="flag" height="200px" width="200px" />
            </div>
            
            <h1> Country :- {props.modalData.name}</h1>
        <h1>Capital :- {props.modalData.capital}</h1>
          
           <div>
               <h1>Language Used</h1>
               <ul>
          {
              props.languages.map((p:any,id:number)=>
          (<li key={id}>{p.name} <span>{"("+p.nativeName+")"}</span></li>)
            )
          }
          </ul>
           </div>
           <div>
               <h1>Currency</h1>
               {
                   props.currency.map((c:any,key:number)=>(
                   <p key={key}>{c.name} , ({c.symbol})</p>
                   ))
               }
           </div>
            <button className="btn"
                onClick={props.handleClose}
            >
                Close
           </button>
        </section>
    </div>)
}
export default Modal;