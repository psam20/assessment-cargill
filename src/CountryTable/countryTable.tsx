import React, { useEffect } from 'react';
import { useSortData } from '../SortConfig/sortConfig'
// import Modal from './Modal'
import './countryTable.css'
import Modal from './Modal';

interface props {
    Countries: Array<Object>,

}
const CountryTable = (props: props) => {

    const { items, requestSort } = useSortData(props.Countries)

    const [show, setShow] = React.useState<boolean>(false)
    const [clickedModalItem,setClickedModalItem]=React.useState<number>(0);
    const [modalData,setModalData]=React.useState<any>({});
    const [lang,setLang] = React.useState([]);
    const [currency,setCurrency]=React.useState([]);
    const showModal=(id:any)=>{
        setShow(true)
        setClickedModalItem(id);
       
    }
    const closeModal=()=>{
        setShow(false)
    }

  

    useEffect(()=>{
         const index=clickedModalItem;
         let modalData;
         let languageData;
         let currencyData;
         if(items.length===0){
            modalData=[];
            languageData=[];
            currencyData=[];
         }
         else{
             modalData=items[index];
             languageData=modalData.languages;
             currencyData=modalData.currencies;
            console.log(languageData)

         }
        setLang(languageData);
        setCurrency(currencyData);
        setModalData(modalData)  
    },[clickedModalItem,items])
  
    return (
        <div className="country-table">
            <div className="header">
                List of Countries
            </div>
            <table cellSpacing="0" className="table">

                <thead>
                    <tr className="tr">
                        
                        <th className="th">
                            <div>
                                <button className="btn" onClick={() => requestSort('name')}>
                                    Country (ASC/DSC)
                          </button>
                            </div>
                        </th>
                        <th className="th">
                            <div>
                                <button className="btn" onClick={() => requestSort('population')}>
                                    Population (ASC/DSC)
                          </button>
                            </div>
                        </th>
                    </tr>
                </thead>
                <tbody>
                    { (items.length===0)?(<tr style={{color:"blue"}}>
                        <td>
                            <h4>
                        NO Country Found...
                        </h4>
                        </td>
                        </tr>):
                        items.map((item, id) => (
                            <tr key={id} className="tr">
                               
                               
                                <td className="td tdh"><div>
                                    <img src={`${item.flag}`} height="20px" alt="flag" width="20px" /> <span onClick={()=>showModal(id)}>{item.name}</span>
                                   
                                    </div></td>
                                <td className="td">{item.population}</td>

                            </tr>
                        ))
                    }
                </tbody>
            </table>
             {(items.length===0)?"":
             <Modal handleClose={closeModal} show={show} currency={currency} languages={lang} modalData={modalData}/>
             }
        </div>
    )
}
export default CountryTable;