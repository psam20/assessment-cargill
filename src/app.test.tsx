import React from 'react';
import {shallow} from 'enzyme';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event'

import App from './App';
import CountryTable from './CountryTable/countryTable';


let data: Array<Object> = []

data = [{
    name: "china",
    population: 320000000,
    region: "asia",
}, {
    name: "india",
    population: 31000000,
    region: "asia",
}]
const filterKey = ["name", "region"]
const filterByKey = (inputArr = data, searchTerm: string) => {
    return inputArr.filter((e: any) => Object.keys(e).some((key: any) =>
        filterKey.includes(key) ? e[key].match(searchTerm) : false
    ))

}


it("renders without crashing", () => {
  expect(shallow(<App />)).toMatchSnapshot();
});

it("Renders Country Table ELement WithOut Crashing", () => {
   expect(shallow(<CountryTable Countries={data} />)).toMatchSnapshot();
})
it('Testing Filter Input ', () => {
    render(<input />)
    userEvent.type(screen.getByRole('textbox'), 'china')
    expect(screen.getByRole('textbox')).toHaveValue('china')
})
describe('testing for Filtering Countries', () => {

    it('filtering Countries By Name', () => {

        const output = [{
            name: "china",
            population: 320000000,
            region: "asia",
        }]
        render(<input />);
        userEvent.type(screen.getByRole('textbox'), 'china');
        
        expect(filterByKey(data,'china')).toEqual(output)
    })

    it('Empty Search Bar', () => {
        expect(filterByKey(data,'')).toEqual(data)
    })
})
