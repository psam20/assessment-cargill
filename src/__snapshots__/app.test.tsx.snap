// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Country Table ELement WithOut Crashing 1`] = `ShallowWrapper {}`;

exports[`renders without crashing 1`] = `ShallowWrapper {}`;
